src/clients/rust/src/time_based_id.rs

```diff
@@ -0,0 +1,314 @@
+use std::cmp::Ordering;
+use std::sync::Mutex;
+use std::time::SystemTime;
+
+/// Generate a TigerBeetle time-based identifier.
+///
+/// This generates `u128` identifiers suitable for the `id` fields
+/// of TigerBeetle `Account`s and `Transaction`s.
+///
+/// [TigerBeetle time-based identifiers][tbid] include a timestamp
+/// component and a random component, are lexicographically sortable,
+/// monotonically increasing, and enable optimizations in TigerBeetle's LSM tree.
+///
+/// [tbid]: https://docs.tigerbeetle.com/coding/data-modeling/#tigerbeetle-time-based-identifiers-recommended
+///
+/// ## System clock warning
+///
+/// Generating these IDs correctly depends on a well-behaved system clock. There
+/// are two cases that will cause degenerate behavior:
+///
+/// - The system clock changes into the past. IDs will be generated
+///   sequentially from the previous ID until the clock catches up with the future timestamp.
+///
+/// - The system time is prior to the Unix epoch. IDs will be generated
+///   sequentially starting at the Unix epoch (plus a base random number).
+//
+// References:
+//
+// - https://github.com/tigerbeetle/tigerbeetle/blob/75f77b8b3280ce2f289cf42ae928945190fe4a2a/src/clients/node/src/index.ts#L161-L191
+// - https://github.com/ulid/spec
+pub fn id() -> u128 {
+    let mut guard = GLOBAL_GENERATOR.lock().expect("global tbid generator");
+    match *guard {
+        None => {
+            *guard = Some(TbidGenerator::new());
+            drop(guard);
+            id()
+        }
+        Some(ref mut generator) => generator.next(),
+    }
+}
```

Comment:
That's for a future improvement, but we might want to have something like

```rust
pub fn ids(count: usize) -> impl ExactSizeIterator<Item=u128>
```

to amortize those mutex lock/unlocks. Or maybe we don't. As I've recently realized, tigerbeetle client library isn't really the right place for id generating function, as ids should be generated end-to-end (on a mobile phone/website, rather than at the API gateway). And at the _end_ you aren't going to generate 8k transfers in one go anyway.

Reply:
Even if the servers themselves shouldn't be generating ids, the id functions have to live and be distributed somewhere. The client libraries seem as good a place as any for now if we don't want everybody reimplementing `id`.